src/ClassDiscovery.php

```diff
@@ -0,0 +1,58 @@
+<?php
+
+namespace Http\Discovery;
+
+/**
+ * Registry that based find results on class existence
+ *
+ * @author David de Boer <david@ddeboer.nl>
+ */
+abstract class ClassDiscovery implements Discovery
```

Comment:
This isn't too much code that complexity worths avoiding duplication. The guess condition logic is a bit different there.